src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,295 @@
+                    case DiscardPattern n: return "_";
+                    case NotNullPattern n: return "{}";
+                    case ConstantPattern n: return $"C:{n.Expression}";
+                    case PositionalPattern n: return $"P:({string.Join(", ", n.Subpatterns.Select(p => p.Pattern))})";
+                    default: return null;
+                }
+            }
+#endif
+        }
+
+        private sealed class Conjunction : AnalyzedNode
+        {
+            public readonly AnalyzedNode Left;
+            public readonly AnalyzedNode Right;
+
+            public override NodeKind Kind => NodeKind.Conjunction;
+
+            public static Conjunction Create(AnalyzedNode left, AnalyzedNode right)
+            {
+                if (left is null || right is null)
+                    return null;
+
+                return new Conjunction(left, right);
+            }
+
+            public Conjunction(AnalyzedNode left, AnalyzedNode right)
+            {
+                Debug.Assert(left != null);
+                Debug.Assert(right != null);
+                Left = left;
+                Right = right;
+            }
+
+            public override bool Contains(ExpressionSyntax e) => Left.Contains(e) || Right.Contains(e);
+
+            public override AnalyzedNode Reduce(Reducer reducer) => reducer.ReduceConjunction(this);
+
+            public override void GetChildren(List<AnalyzedNode> nodes)
+            {
+                Left.GetChildren(nodes);
+                Right.GetChildren(nodes);
+            }
+
+            public override PatternSyntax AsPatternSyntax() => AsPatternSyntax(out _);
+
+            private PatternSyntax AsPatternSyntax(out List<AnalyzedNode> children)
+            {
+                children = new List<AnalyzedNode>();
+                GetChildren(children);
+                return RecursivePattern(
+                    children.OfType<TypePattern>().SingleOrDefault()?.Type,
+                    children.OfType<PositionalPattern>().SingleOrDefault()?.AsPositionalPatternClauseSyntax(),
+                    PropertyPatternClause(SeparatedList(
+                        children.OfType<PatternMatch>().Select(match => match.AsSubpatternSyntax()))),
+                    children.OfType<VarPattern>().SingleOrDefault()?.AsVariableDesignationSyntax());
+            }
+
+            public override ExpressionSyntax AsExpressionSyntax()
+                => BinaryExpression(SyntaxKind.LogicalAndExpression,
+                    Left.AsExpressionSyntax(), Right.AsExpressionSyntax());
+
+            public SyntaxNode AsCasePatternSwitchLabelSyntax()
+                => CasePatternSwitchLabel(AsPatternSyntax(out var children),
+                    AsWhenClause(children.OfType<Evaluation>().Select(e => e.Expression).AggregateOrDefault(
+                        (left, right) => BinaryExpression(SyntaxKind.LogicalAndExpression, left, right))),
+                    Token(SyntaxKind.ColonToken));
```

Comment:
dumb question.  why is AsCasePatternSwitchLabelSyntax not the same as just calling AsPatternSyntax and wrapping that with a CasePatternSwitchLabel?

Reply:
either:
1. there is a good reason for it to be different.  but we need to doc taht in the code.
2. they shouldn't both need to exist, in which case, we can remove one of these.